drivers/gpio/gpio_cy8c95xx.c

```diff
@@ -0,0 +1,347 @@
+			      ((out >> 12) & 0x0F) };
+
+	int rc = i2c_burst_write(drv_data->i2c_master, cfg->i2c_slave_addr,
+				 CY8C95XX_REG_OUTPUT_DATA0, tx_buf, 3);
+
+	if (rc == 0) {
+		*outp = out;
+	}
+	k_sem_give(&drv_data->lock);
+
+	LOG_DBG("write %04x msk %04x val %04x => %04x: %d", orig_out, mask, value, out, rc);
+
+	return rc;
+}
+
+static int port_set_masked(const struct device *dev,
+			   gpio_port_pins_t mask,
+			   gpio_port_value_t value)
+{
+	return port_write(dev, mask, value, 0);
+}
+
+static int port_set_bits(const struct device *dev,
+			 gpio_port_pins_t pins)
+{
+	return port_write(dev, pins, pins, 0);
+}
+
+static int port_clear_bits(const struct device *dev,
+			   gpio_port_pins_t pins)
+{
+	return port_write(dev, pins, 0, 0);
+}
+
+static int port_toggle_bits(const struct device *dev,
+			    gpio_port_pins_t pins)
+{
+	return port_write(dev, 0, 0, pins);
+}
+
+static int pin_interrupt_configure(const struct device *dev,
+				   gpio_pin_t pin,
+				   enum gpio_int_mode mode,
+				   enum gpio_int_trig trig)
+{
+	return -ENOTSUP;
+}
+
+/**
+ * @brief Initialization function of CY8C95XX
+ *
+ * @param dev Device struct
+ * @return 0 if successful, failed otherwise.
+ */
+static int cy8c95xx_init(const struct device *dev)
+{
+	const struct cy8c95xx_config *cfg = dev->config;
+	struct cy8c95xx_drv_data *drv_data = dev->data;
+	int rc;
+	uint8_t data = 0;
+
+	drv_data->i2c_master = device_get_binding(cfg->i2c_master_dev_name);
+	if (!drv_data->i2c_master) {
+		LOG_ERR("%s: no bus %s", dev->name,
+			cfg->i2c_master_dev_name);
+		rc = -EINVAL;
+		goto out;
+	}
+
+	i2c_reg_read_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+			  CY8C95XX_REG_ID, &data);
+	LOG_DBG("cy8c95xx device ID %02X", data & 0xF0);
+	if ((data & 0xF0) != 0x20) {
+		LOG_WRN("driver only support [0-2] ports operations");
+	}
+
+	/* port 0 configuration */
+	i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+			   CY8C95XX_REG_PORT_SELECT, 0x00);
+	i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+			   CY8C95XX_REG_PULL_UP, 0xFF);
+	/* port 1 configuration */
+	i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+			   CY8C95XX_REG_PORT_SELECT, 0x01);
+	i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+			   CY8C95XX_REG_PULL_UP, 0xFF);
+	/* port 2 configuration */
+	i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+			   CY8C95XX_REG_PORT_SELECT, 0x02);
+	i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+			   CY8C95XX_REG_PULL_UP, 0xFF);
+
+
+	/* Reset state mediated by initial configuration */
+	drv_data->pin_state = (struct cy8c95xx_pin_state) {
+		.dir = (ALL_PINS & DT_INST_PROP(0, init_dir)),
+		.data = (ALL_PINS & DT_INST_PROP(0, init_data)),
+	};
+	rc = write_pin_state(cfg, drv_data, &drv_data->pin_state);
+out:
+	if (rc != 0) {
+		LOG_ERR("%s init failed: %d", dev->name, rc);
+	} else {
+		LOG_INF("%s init ok", dev->name);
+	}
+	k_sem_give(&drv_data->lock);
```

Comment:
was there a k_sem_take before?

Reply:
the sem's initial value will is 0, so we don't need to take it.